File: field_force/field_force/page/store_visit_details_report/store_visit_details_report.py
import frappe
import json

from field_force.field_force.page.utils import generate_excel_and_download, get_time_in_12_hour_format, get_datetime_with_12_hour_format
from field_force.field_force.report.utils import set_image_url, get_site_directory_path, set_link_to_doc


@frappe.whitelist()
def get_store_visit_data(filters=None):
    columns = get_columns()
    data = get_absolute_data(filters)
    return data, columns

def get_absolute_data(filters, export=False):
    query_result = get_query_data(filters)
    site_directory = get_site_directory_path()

    for index, store_visit in enumerate(query_result):
        set_image_url(store_visit, site_directory)

        server_time = get_time_in_12_hour_format(store_visit.server_time)
        device_time = get_time_in_12_hour_format(store_visit.device_time)
        # server_time = str(store_visit.server_time).split('.')[0] \
        #     if '.' in str(store_visit.server_time) else store_visit.server_time
        # device_time = str(store_visit.device_time).split('.')[0] \
        #     if '.' in str(store_visit.device_time) else store_visit.device_time

        if not export:
            set_link_to_doc(store_visit, 'name', 'store-visit')
            set_link_to_doc(store_visit, 'customer', 'customer')
            set_link_to_doc(store_visit, 'sales_person', 'sales-person')

            # store_visit.server_date = f"{store_visit.server_date}<br>{server_time}"
            store_visit.server_time = server_time
            store_visit.device_date = f"{store_visit.device_date}<br>{device_time}"
        else:
            # store_visit.user = store_visit.user_fullname or store_visit.user
            # store_visit.server_date = f"{store_visit.server_date}\n{server_time}"
            store_visit.server_time = server_time
            store_visit.device_date = f"{store_visit.device_date}\n{device_time}"

        store_visit.cheated = 'Yes' if store_visit.cheated else 'No'
        store_visit['sl'] = index + 1

    return query_result

def get_query_data(filters):
    try:
        filters = json.loads(filters)
    except:
        pass
    conditions = get_conditions(filters)

    query_string = '''select store_visit.name, store_visit.sales_person, store_visit.customer, store_visit.details,
                    store_visit.image, store_visit.contact_number, store_visit.server_date, store_visit.server_time,
                    store_visit.device_date, store_visit.device_time, store_visit.latitude, store_visit.longitude,
                    store_visit.device_model, store_visit.customer_address, store_visit.cheated from `tabStore Visit` 
                    store_visit where %s order by store_visit.server_date desc, store_visit.server_time desc''' % conditions

    query_result = frappe.db.sql(query_string, as_dict=1, debug=0)
    return query_result


def get_conditions(filters):
    from_date = filters.get('from_date')
    to_date = filters.get('to_date')
    sales_person = filters.get('sales_person')
    customer = filters.get('customer')

    conditions = []

    if from_date:
        conditions.append('store_visit.server_date >= "%s"' % from_date)
    if to_date:
        conditions.append('store_visit.server_date <= "%s"' % to_date)
    if sales_person:
        conditions.append('store_visit.sales_person = "%s"' % sales_person)
    if customer:
        conditions.append('store_visit.customer = "%s"' % customer)

    return " and ".join(conditions)

def get_columns():
    columns =  [
        {'fieldname': 'sl', 'label': 'SL', 'expwidth': 5, 'export': False, 'width': 30},
        {'fieldname': 'server_date', 'label': 'Date', 'expwidth': 13, 'width': 90},
        {'fieldname': 'server_time', 'label': 'Time', 'expwidth': 13, 'width': 70},
        {'fieldname': 'name', 'label': 'ID', 'expwidth': 20},
        {'fieldname': 'customer', 'label': 'Customer', 'expwidth': 15, 'width':150},
        # {'fieldname': 'customer_address', 'label': 'Address', 'expwidth': 15},
        # {'fieldname': 'contact_number', 'label': 'Contact', 'expwidth': 15},
        {'fieldname': 'details', 'label': 'Details', 'expwidth': 15, 'width':100},
        {'fieldname': 'sales_person', 'label': 'Sales Person', 'expwidth': 20},
        {'fieldname': 'device_date', 'label': 'Device Date Time', 'expwidth': 15, 'width': 80},
        {'fieldname': 'cheated', 'label': 'Cheated', 'fieldtype': 'Data', 'expwidth': 15, 'width': 20},
        # {'fieldname': 'latitude', 'label': 'Latitude', 'fieldtype': 'Data', 'expwidth': 15},
        # {'fieldname': 'longitude', 'label': 'Longitude', 'fieldtype': 'Data', 'expwidth': 15},
        # {'fieldname': 'device_model', 'label': 'Device Model', 'fieldtype': 'Data', 'expwidth': 15},
        {'fieldname': 'image', 'label': 'Image', 'fieldtype': 'Image', 'expwidth': 15, 'export': False},
    ]
    return columns

@frappe.whitelist()
def export_file(**filters):
    columns = get_columns()
    data = get_export_data(filters)
    file_name = 'Store_Visit_Details_Report.xlsx'
    generate_excel_and_download(columns, data, file_name, height=30)

def get_export_data(filters):
    query_result = get_absolute_data(filters, export=True)
    return query_result

File: field_force/field_force/page/store_visit_details_report/store_visit_details_report.js
frappe.pages['store-visit-details-report'].on_page_load = function(wrapper) {
    frappe.require("assets/field_force/css/page_modal.css");
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Store Visit Details Report',
        single_column: true
    });
    $('.page-body').css('background', '#FFFFFF');
    new StoreVisitDetailsReport(page);
};

class StoreVisitDetailsReport {
    constructor(page) {
        this.page = page;
        this.make_form();
        this.make_menu();
        this.initialize_modal();
        // this.open_image_modal()
    }

    make_menu = () => {
        this.page.add_action_icon("refresh", () => {
            this.fetch_and_render();
        });
        this.page.add_menu_item("Export", ()=>{
            this.export_excel()
        })
    }
    make_form = () => {
        this.form = new frappe.ui.FieldGroup({
            fields: [
                {

                    fieldname: 'from_date',
                    label: __('From Date'),
                    fieldtype: 'Date',
                    default: frappe.datetime.get_today(),
                    change: () => this.fetch_and_render(),
                    reqd: 1,
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'to_date',
                    label: __('To Date'),
                    fieldtype: 'Date',
                    default: frappe.datetime.get_today(),
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'customer',
                    label: __('Customer'),
                    fieldtype: 'Link',
                    options: 'Customer',
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'sales_person',
                    label: __('Sales Person'),
                    fieldtype: 'Link',
                    options: 'Sales Person',
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Section Break'
                },
                {
                    fieldtype: 'HTML',
                    fieldname: 'preview'
                }
            ],
            body: this.page.body
        });
        this.form.make();
    }
    initialize_modal = () => {
        let html = '<div id="modal_section" class="modal_">\n' +
            '  <div class="modal-image">\n' +
            '  <img class="modal-content_" id="img01" alt="img">\n' +
            '  <span class="close" onclick="document.getElementById(\'modal_section\').style.display=\'none\'">&times;</span>\n' +
            '  </div>\n' +
            '</div>'
        // this.form.body.append(html);
        $('.main-section').append(html)
    }
    fetch_and_render = () => {
        let {from_date, to_date, sales_person, customer} = this.form.get_values();
        if (!from_date) {
            this.form.get_field('preview').html('');
            return;
        }
        this.form.get_field('preview').html(`
        	<div class="text-muted margin-top">
        		${__("Fetching...")}
        	</div>
        `);
        frappe.call('field_force.field_force.page.store_visit_details_report.store_visit_details_report.get_store_visit_data', {
            filters: {
                from_date: from_date,
                to_date: to_date,
                sales_person: sales_person,
                customer: customer
            },
            freeze: true
        }).then(r => {
            let diff = r.message[0];
            let fields = r.message[1];
            this.render(diff, fields);
        });
    }

    render = (diff, fields) => {
        let table_header = this.table_header(fields);
        let table_body = this.table_body(diff, fields);
        this.form.get_field('preview').html(`<table class="table table-bordered" id="export_excel">${table_header}${table_body}</table>`);
    }

    table_header = (headers) => {
        let table_header = `<thead><tr>`;

        headers.forEach(function (field, index){
            table_header += `<th scope="col" width="${field.width}">${field.label || ''}</th>`;
        })

        table_header += `</tr>`;
        return table_header;
    }

    table_body = (diff, fields) => {
        var html = `<tbody>`;

        diff.forEach(function (data, index) {
            html += `<tr>`;

            fields.forEach(function (field, index){
               html += get_absolute_format_and_html(field, data[field.fieldname], field.width);
            })

            html += `</tr>`;
        })

        html += `</tbody>`;
        return html
    }
    export_excel = () => {
        let {from_date, to_date, customer, sales_person} = this.form.get_values();
        let url = `/api/method/field_force.field_force.page.store_visit_details_report.store_visit_details_report.export_file`;
        url += `?from_date=${from_date||''}&to_date=${to_date||''}&customer=${customer||''}&sales_person=${sales_person||''}`;
        window.open(url, '_blank');
    }
}


function get_absolute_format_and_html(field, value, width=50){
    if (field.fieldtype === "Image"){
        return get_image_html(value);
    }
    else if (field.fieldtype === "Currency"){
        return get_currency_format(value);
    }
    else if (field.fieldtype ==="Data") {
        return `<td width="${width}px">${value || ''}</td>`;
    }
    else {
        return `<td width="${width}px">${value || ''}</td>`;
    }
}

function get_image_html(image_url) {
    return `
        <td style="height:100px; width:120px;">
            <a href="#">
                <img style="height:100%; width:100%" src="${image_url}" onclick="(
                    function(e){
                        document.getElementById(\'modal_section\').style.display=\'block\';
                        var nAgt = navigator.userAgent;
                        if (nAgt.indexOf('Safari') !== -1) {
                            document.getElementById(\'img01\').src=e.target.currentSrc;
                        }else{
                            document.getElementById(\'img01\').src=e.path[0].currentSrc;
                        }
                        return false;
                    }
                )(arguments[0]);
                return false;">
            </a>
        </td>`;
}

function get_currency_format(value){
    const company = frappe.defaults.get_user_default("company");
    const currency = frappe.get_doc(":Company", company).default_currency;
    let value_in_currency = format_currency(value, currency);
    return `<td>${value_in_currency || ''}</td>`
}